compiler/src/iree/compiler/Codegen/Common/CleanupBufferAllocViewPass.cpp

```diff
@@ -101,6 +107,76 @@ struct FoldReshapeIntoInterfaceTensorLoad : OpRewritePattern<TensorReshapeOp> {
   }
 };
 
+/// Folds tensor.expand/collapse_shape into the source
+/// hal.interface.binding.subspan.
+///
+/// For example, this matches the following pattern:
+///
+///   %subspan = hal.interface.binding.subspan ... :
+///       !flow.dispatch.tensor<writeonly:tensor<3x3x1x96xf32>>
+///   %0 = linalg.tensor_reshape %tensor [
+///         affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>
+///       ] : tensor<864xf32> into tensor<3x3x1x96xf32>
+///   %tensor = flow.dispatch.tensor.store %0, %subspan :
+///       !flow.dispatch.tensor<writeonly:tensor<3x3x1x96xf32>> ->
+///       tensor<3x3x1x96xf32>
+///
+/// And turns it into:
+///
+///   %subspan = hal.interface.binding.subspan ... :
+///       !flow.dispatch.tensor<writeonly:tensor<864xf32>>
+///   %0 = flow.dispatch.tensor.store %tensor, %subspan :
+///       !flow.dispatch.tensor<writeonly:tensor<864xf32>> -> tensor<864xf32>
+struct FoldReshapeIntoInterfaceTensorStore
+    : OpRewritePattern<IREE::Flow::DispatchTensorStoreOp> {
+  using OpRewritePattern<IREE::Flow::DispatchTensorStoreOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(IREE::Flow::DispatchTensorStoreOp storeOp,
+                                PatternRewriter &rewriter) const override {
+    // Make sure we are storing the full incoming subspan. Otherwise we cannot
+    // simply adjust the subspan's resultant type later.
+    if (!storeOp.offsets().empty() || !storeOp.sizes().empty() ||
+        !storeOp.strides().empty())
+      return failure();
+
+    auto reshapeOp = storeOp.getValue().getDefiningOp();
+    if (!isa<tensor::CollapseShapeOp, tensor::ExpandShapeOp>(reshapeOp))
+      return failure();
+
+    // Dynamic shapes are currently unsupported.
+    Optional<Value> reshapeSrc =
+        isa<tensor::CollapseShapeOp>(reshapeOp)
+            ? getStaticReshapeOpSrc<tensor::CollapseShapeOp>(
+                  cast<tensor::CollapseShapeOp>(reshapeOp))
+            : getStaticReshapeOpSrc<tensor::ExpandShapeOp>(
+                  cast<tensor::ExpandShapeOp>(reshapeOp));
+    if (!reshapeSrc) return failure();
+
+    auto subspanOp =
+        storeOp.getTarget()
+            .template getDefiningOp<IREE::HAL::InterfaceBindingSubspanOp>();
+    if (!subspanOp) return failure();
+    assert(subspanOp.getDynamicDims().empty());
+
+    auto tensorAccess = subspanOp.getType()
+                            .template cast<IREE::Flow::DispatchTensorType>()
+                            .getAccess();
+    auto newSubspanType = IREE::Flow::DispatchTensorType::get(
+        tensorAccess, reshapeSrc->getType());
+
+    Value newSubspanOp = rewriter.create<IREE::HAL::InterfaceBindingSubspanOp>(
```

Comment:
You want to insert the interface binding op at the same place as the original op. This can lead to some strange compilation effects otherwise.

Reply:
I see that this affects bufferization, although I'm now seeing this for my particular case after bufferization
```
%1 = hal.interface.binding.subspan set(0) binding(2) type(storage_buffer) alignment(64) offset(%c0) : memref<2x32x16384xf32, #hal.descriptor_type<storage_buffer>>
%3 = hal.interface.binding.subspan set(0) binding(2) type(storage_buffer) alignment(64) offset(%c0) : memref<2x32x16384xf32, #hal.descriptor_type<storage_buffer>>
...
linalg.generic {indexing_maps = [#map9, #map9], iterator_types = ["parallel", "parallel", "parallel"]} ins(%3 : memref<2x32x16384xf32, #hal.descriptor_type<storage_buffer>>) outs(%1 : memref<2x32x16384xf32, #hal.descriptor_type<storage_buffer>>) {
^bb0(%in: f32, %out: f32):
  linalg.yield %in : f32
}
```
Should I rely on CSE + canonicalization patterns to avoid this or somehow handle it here?

Reply:
I dont know if those two interface bindings CSE. They should, but they dont cause they are tagged with memory effects. Maybe we can mark these ops as non-side effecting. Then CSE will do what you need.

Reply:
That worked, I'll make that a separate PR then.

Reply:
Use RAII

```
{
   OpBuilder::InsertionGuard guard(rewriter);
   rewriter.setInsertionPointAfter(subspanOp);
   rewriter.create<IREE::HAL::INterfaceBindingSubspanOp>
}
```

Then you can drop the subsequent insertion point setting. That can be a foot gun.